CRISPResso2/CRISPRessoCORE.py

```diff
@@ -501,6 +862,488 @@ def custom_encoder(obj):
+            fastq2_seq = CRISPRessoShared.reverse_complement(fastq2_handle.readline().strip())
+            fastq2_plus = fastq2_handle.readline()
+            fastq2_qual = fastq2_handle.readline().strip()[::-1]
+
+            if (N_TOT_READS % 10000 == 0):
+                info("Processing reads; N_TOT_READS: %d N_COMPUTED_ALN: %d N_CACHED_ALN: %d N_COMPUTED_NOTALN: %d N_CACHED_NOTALN: %d"%(N_TOT_READS, N_COMPUTED_ALN, N_CACHED_ALN, N_COMPUTED_NOTALN, N_CACHED_NOTALN))
+
+            N_TOT_READS+=1
+            #if the sequence has been seen and can't be aligned, skip it
+            #cache the sequence of both r1 and r2 sequences as lookup_fastq_seq
+            lookup_fastq_seq = fastq1_seq + "+" + fastq2_seq
+            if lookup_fastq_seq in not_aln:
+                N_CACHED_NOTALN += 1
+                if args.fastq_output:
+                    crispresso2_annotation = " ALN=NA" +\
+                        " ALN_SCORES=" + ('&'.join([str(x) for x in not_aln[lookup_fastq_seq]['aln_scores']])) +\
+                        " ALN_DETAILS=" + ('&'.join([','.join([str(y) for y in x]) for x in not_aln[lookup_fastq_seq]['ref_aln_details']]))
+                    fastq_write_out_handle.write(f"{fastq1_id.strip()}\n{not_aln[lookup_fastq_seq]['ref_aln_details'][0][1].replace('-', '')}\n+{crispresso2_annotation}\n{not_aln[lookup_fastq_seq]['ref_aln_details'][0][4]}\n")
+                continue
+            # if the sequence is already associated with a variant in the variant cache, pull it out
+            if lookup_fastq_seq in variantCache:
+                N_CACHED_ALN += 1
+                variantCache[lookup_fastq_seq]['count'] += 1
+                if args.fastq_output:
+                    new_variant = variantCache[lookup_fastq_seq]
+
+                    ins_inds = []
+                    del_inds = []
+                    sub_inds = []
+                    edit_strings = []
+
+                    for idx, best_match_name in enumerate(new_variant['aln_ref_names']):
+                        payload=new_variant['variant_'+best_match_name]
+
+                        del_inds.append([str(x[0][0])+"("+str(x[1])+")" for x in zip(payload['deletion_coordinates'], payload['deletion_sizes'])])
+
+                        ins_vals = []
+                        for ins_coord,ins_size in zip(payload['insertion_coordinates'],payload['insertion_sizes']):
+                            ins_start = payload['ref_positions'].index(ins_coord[0])
+                            ins_vals.append(payload['aln_seq'][ins_start+1:ins_start+1+ins_size])
+                        ins_inds.append([str(x[0][0])+"("+str(x[1])+"+"+x[2]+")" for x in zip(payload['insertion_coordinates'], payload['insertion_sizes'], ins_vals)])
+
+                        sub_inds.append(payload['substitution_positions'])
+                        edit_strings.append('D'+str(int(payload['deletion_n']))+';I'+str(int(payload['insertion_n']))+';S'+str(int(payload['substitution_n'])))
```

Comment:
These are only needed `if args.fastq_output`, right? If this is correct, it would improve performance if we place this inside the `if` block. This would also apply down in lines 1241-1254, and I think this would also apply to the other `process_*` functions. Do you think that would be better as a separate PR?

Also, not necessary, but a good "quality of life" improvement may be to extract this into a function that calculates these metrics and returns `crispresso2_annotation` (and then we could use it in the various other `process_*` functions).